### PR TITLE
Add 'yq' option in traverse (#15)

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: git_machete/cmd.py [hunk 1]
```diff
@@ -283,7 +283,7 @@ def up(b, prompt_if_inferred):
         u = infer_upstream(b)
         if u:
             if prompt_if_inferred:
-                if ask_if("Branch '%s' not found in the tree of branch dependencies; rebase onto the inferred upstream '%s'? [y/n] " % (b, u)):
```
Comment (pasuder) [inline git_machete/cmd.py:286]: I would like to ask you to start limiting length to 80 chars. It is tricky to review on GitHub without vertical scrollbar which is on the bottom.

File: git_machete/cmd.py [hunk 2]
```diff
@@ -901,61 +901,71 @@ def print_new_line(new_status):
...
                     del annotations[b]
                 save_definition_file()
+                if ans == 'yq':
+                    raise StopTraversal
                 # No need to flush caches since nothing changed in commit/branch structure (only machete-specific changes happened).
-                continue  # No need to sync branch `b` with remote since it just got removed from the tree of dependencies.
+                continue  # No need to sync branch 'b' with remote since it just got removed from the tree of dependencies.
             elif ans in ('q', 'quit'):
-                return
+                raise StopTraversal
```
Comment (pasuder) [inline git_machete/cmd.py]: It seems to be a change of logic. Could you split it into separate commit, please?

**Review by pasuder — CHANGES_REQUESTED** [review]
I could see here two changes:

- related to overall format and some renaming variables along with adding option `yq`
- change of `return` logic to `raise StopTraversal` - it is not mentioned in commit message - I would like to see it in separate commit.

It would be great to start using limit of 80 chars (or 120) to have possibility to review it without horizontal scrollbar usage.

File: git_machete/cmd.py (hunk 1 above)
Comment (PawelLipski) [reply git_machete/cmd.py:286]: Done, I limited all line lengths in `traverse` method.

File: git_machete/cmd.py (hunk 2 above)
Comment (PawelLipski) [reply git_machete/cmd.py]: I tidied up the raise/return logic in `traverse` and dependent functions, now return is used in `traverse` and `raise` in dependent functions, hope it's clearer now.

**Comment by PawelLipski** [comment]
@pasuder fixes applied on a separate commit

File: git_machete/cmd.py (hunk 1 above)
Comment (pasuder) [reply git_machete/cmd.py:286]: Nice, thx!

File: git_machete/cmd.py [hunk 3]
```diff
@@ -856,17 +856,123 @@ def __init__(self):
...
+            flush()
+        elif can_pick_other_remote and ans in ('o', 'other'):
+            pick_remote(b)
+        elif ans in ('q', 'quit'):
+            raise StopTraversal
+        return
+
+    new_s = get_relation_to_remote_counterpart(b, rb)
+    if new_s == IN_SYNC_WITH_REMOTE:
```
Comment (pasuder) [inline git_machete/cmd.py]: Please decompose those if-elif into small functions.

**Comment by pasuder** [comment]
Those if-elif-else aka _forest of Elves_ should be decomposed into functions and instead of if-elif, dict with names and associated functions should be used.

File: git_machete/cmd.py (hunk 3 above)
Comment (PawelLipski) [reply git_machete/cmd.py]: Done

**Comment by PawelLipski** [comment]
@pasuder: refactored `handle_untracked_branch` in a similar manner to what you suggested

File: git_machete/cmd.py [hunk 4]
```diff
@@ -856,17 +856,123 @@ def __init__(self):
         pass
 
 
+def flush():
+    global branch_defs_by_sha_in_reflog, initial_log_shas_cached, remaining_log_shas_cached, remote_tracking_branches_cached, sha_by_revision_cached
```
Comment (pasuder) [inline git_machete/cmd.py:860]: This should be handled in a bit different way. It looks very bash'y way.

File: git_machete/cmd.py [hunk 5]
```diff
@@ -856,17 +856,123 @@ def __init__(self):
...
+            if ans == 'yq':
+                raise StopTraversal
+            flush()
+        elif can_pick_other_remote and ans in ('o', 'other'):
+            pick_remote(b)
+        elif ans in ('q', 'quit'):
+            raise StopTraversal
+        return
+
+    message = {
```
Comment (pasuder) [inline git_machete/cmd.py:904]: Nice!

**Review by pasuder — DISMISSED** [review]
Looks good, one comment for future task about those globals.

File: git_machete/cmd.py (hunk 4 above)
Comment (PawelLipski) [reply git_machete/cmd.py:860]: Yes, I'm aware of how much mess with globals I made in this script :D there's a task to tidy up all that (#27)
